subm/src/sheetAccess.js

```diff
@@ -0,0 +1,89 @@
+const range = n => [...Array(n).keys()];
+const takeWhile = (xs, predicate) =>
+  xs.slice(
+    0,
+    xs.findIndex(x => !predicate(x)),
+  );
+
+/**
+ * @param {GoogleSpreadsheetWorksheet} sheet
+ * @param {string | number} key
+ */
+const lookup = async (sheet, key) => {
+  // load primary key column
+  await sheet.loadCells({
+    startColumnIndex: 0,
+    endColumnIndex: 1,
+  });
+
+  const destIndex = range(sheet.rowCount).find(rowIndex => {
+    if (rowIndex === 0) return false; // header row
+    const { value } = sheet.getCell(rowIndex, 0);
+    return key === value || value === null;
```

Comment:
What is the purpose of `value === null` here? I am concerned that this will return an empty row, and that simultaneous `upsert` calls will pick the same empty row, with a last one win insertion.

Reply:
when upsert inserts, the row index is not used. it just calls addRow. But still, the `value === null` does seem left over from an earlier design.

Reply:
ugh. that was a bit of a pain to work through. had to fall back from `range(n).find(...)` to a `for` loop.